test/ARCHITECTURE.md

```diff
@@ -0,0 +1,259 @@
+usually based on a distribution's cloud image customized with:
+
+* A well-known password for the admin/root user
+* Test SSH keys for access
+* Test packages required to test Cockpit
+* A build chroot with Cockpit's build dependencies to build the to be tested
+  Cockpit source inside the virtual machine offline. This allows a developer on
+  Fedora to easily prepare a Debian test image without having to install Debian
+  build tools.
+* Disabling system services which interfere with testing
+
+To automate user actions on the web page, Cockpit is tested in a browser
+controlled using the [Chrome DevTools Protocol](https://chromedevtools.github.io/devtools-protocol/)
+(CDP) which is supported by Firefox and Chromium based browsers.
+
+The Python test framework in [test/common](./common) is responsible for setting up the
+test environment, running tests and reporting of the test output.
+
+Diagram of the interaction of Browser/Machine/CDP/Test Framework.
+```mermaid
+graph TD;
+    id[Test Framework] <-->|CDP| Browser;
+    Browser <-->|HTTPS/WebSocket| Cockpit;
+    id1[Virtual Machine] <-->|HTTP/DBus/OS API's| Cockpit;
+    id[Test Framework] <-->|SSH| id1[Virtual Machine];
+```
+
+## Integration Tests
+
+Cockpit's tests can be run via three different entry points:
+
+* `test/verify/check-$page` - run a single or multiple unit test(s)
+* `test/common/run-tests` - run tests through our test scheduler (retries, tracks naughties, parallelism)
+* `test/run` - CI entry point
+
+We will start with how a single integration test is run and then explore the test
+scheduler and CI setup.
+
+### Test runtime
+
+The base of a Cockpit integration test looks as following:
+
+```python
+class TestApps(testlib.MachineCase):
+    def testBasic(self):
+        self.machine.execute("rm /etc/machine-id")
+        self.login_and_go("/system")
+        self.browser.wait_text("#system_machine_id", "")
+
+if __name__ == '__main__':
+    testlib.test_main()
+```
+
+In Cockpit there are two types of tests, 'destructive" and "nondestructive".
+tests. Destructive tests do something to the virtual machine which makes it
+unable to run another test afterwards or requires another virtual machine for
+testing. The test above is a `destructive` test which is the default, a non
+`destructive` test makes sure any destructive action is restored after the test
+has run as can be seen below. Nondestructive tests where introduced to be able
+to run them on a single running machine, such as the Testing Farm or a custom VM.
+They also speed up testing, as rebooting and shutting down a machine for every tests incurs a
+significant penalty of ~ 10-30 seconds per test.
+
+```python
+@testlib.nondestructive
+class TestApps(testlib.MachineCase):
+    def testBasic(self):
+        self.restore_file("/usr/share/cockpit/apps")
+        self.machine.execute("rm /etc/machine-id")
+        self.login_and_go("/system")
+        self.browser.wait_text("#system_machine_id", "")
```

Comment:
This is very incoherent. restore_file() does not work with a directory, it rms a different file than it (tries to) restore, it checks /system but is called TestApps, etc.